File: src/backend/ai/resilienceScore.js
// backend/ai/resilienceScore.js

/**
 * Menghitung Resilience Score user (0–100)
 * Analoginya seperti "credit score" tapi untuk melindungi user, bukan bank
 */

function calculateResilienceScore(userData) {
    const {
        totalSavings,           // total tabungan (Rp)
        monthlyExpenses,        // rata-rata pengeluaran per bulan (Rp)
        monthlyIncome,          // income per bulan (Rp)
        totalMonthlyDebtPayment,// total cicilan per bulan (BNPL + dll)
        hasInsurance,           // boolean
        incomeHistory           // array income 3 bulan terakhir, contoh: [4800000, 5000000, 5200000]
    } = userData;

  // ─────────────────────────────────────────
  // PILAR 1: Savings Runway (bobot 35%)
  // "Berapa bulan kamu bisa survive tanpa income?"
  // ─────────────────────────────────────────
    const savingsRunwayMonths = totalSavings / monthlyExpenses;
    // 0 bulan = 0 poin, 6 bulan atau lebih = 100 poin
    const savingsScore = Math.min((savingsRunwayMonths / 6) * 100, 100);

  // ─────────────────────────────────────────
  // PILAR 2: Income Stability (bobot 25%)
  // "Seberapa konsisten income-mu?"
  // ─────────────────────────────────────────
    let stabilityScore = 100; // default jika hanya 1 data
    if (incomeHistory && incomeHistory.length > 1) {
        const mean = incomeHistory.reduce((a, b) => a + b, 0) / incomeHistory.length;
        const variance = incomeHistory.reduce((sum, x) => sum + Math.pow(x - mean, 2), 0) / incomeHistory.length;
        const stdDev = Math.sqrt(variance);
        const coefficientOfVariation = stdDev / mean; // 0 = sangat stabil, >1 = sangat tidak stabil
        stabilityScore = Math.max(0, 100 - coefficientOfVariation * 200);
    }

  // ─────────────────────────────────────────
  // PILAR 3: Debt Burden (bobot 25%)
  // "Seberapa berat beban hutangmu?"
  // Debt-to-Income ratio: <30% sehat, >50% bahaya
  // ─────────────────────────────────────────
    const debtToIncomeRatio = totalMonthlyDebtPayment / monthlyIncome;
  // 0% = 100 poin, 50% atau lebih = 0 poin
    const debtScore = Math.max(0, 100 - (debtToIncomeRatio / 0.5) * 100);

  // ─────────────────────────────────────────
  // PILAR 4: Insurance Coverage (bobot 15%)
  // Sederhana untuk hackathon: punya = 100, tidak = 0
  // ─────────────────────────────────────────
    const insuranceScore = hasInsurance ? 100 : 0;

  // ─────────────────────────────────────────
  // FINAL SCORE (weighted average)
  // ─────────────────────────────────────────
    const finalScore = (
        savingsScore    * 0.35 +
        stabilityScore  * 0.25 +
        debtScore       * 0.25 +
        insuranceScore  * 0.15
    );

    const score = Math.round(finalScore);

  // Generate warning message untuk UI
    const warningMessage = generateWarning(score, savingsRunwayMonths, debtToIncomeRatio);

    return {
        score,
        savingsRunwayMonths: parseFloat(savingsRunwayMonths.toFixed(1)),
        debtToIncomeRatio: parseFloat((debtToIncomeRatio * 100).toFixed(1)), // dalam %
        breakdown: {
        savingsScore:   Math.round(savingsScore),
        stabilityScore: Math.round(stabilityScore),
        debtScore:      Math.round(debtScore),
        insuranceScore: Math.round(insuranceScore),
        },
        warningMessage,
        riskLevel: getRiskLevel(score),
    };
}

function getRiskLevel(score) {
    if (score >= 75) return 'AMAN';
    if (score >= 50) return 'WASPADA';
    if (score >= 25) return 'BERISIKO';
    return 'KRITIS';
}

function generateWarning(score, savingsRunwayMonths, dtiRatio) {
    if (savingsRunwayMonths < 1) {
        return `Tabunganmu hanya cukup untuk ${(savingsRunwayMonths * 30).toFixed(0)} hari. Ini darurat.`;
    }
    if (dtiRatio > 0.4) {
        return `${(dtiRatio * 100).toFixed(0)}% income-mu habis untuk bayar hutang. Di atas batas aman.`;
    }
    if (score < 50) {
        return `Kamu bisa bertahan ${savingsRunwayMonths} bulan tanpa income. Perlu tindakan segera.`;
    }
    return `Kondisi keuanganmu cukup stabil. Kamu bisa bertahan ${savingsRunwayMonths} bulan tanpa income.`;
}

module.exports = { calculateResilienceScore };

File: src/backend/ai/resilienceScore.py
# ai/resilience_score.py
import math

def calculate_resilience_score(user: dict) -> dict:
    total_savings              = user["total_savings"]
    monthly_expenses           = user["monthly_expenses"]
    monthly_income             = user["monthly_income"]
    total_monthly_debt_payment = user["total_monthly_debt_payment"]
    has_insurance              = user["has_insurance"]
    income_history             = user.get("income_history", [monthly_income])

    savings_runway = total_savings / monthly_expenses
    savings_score  = min((savings_runway / 6) * 100, 100)

    if len(income_history) > 1:
        mean    = sum(income_history) / len(income_history)
        std_dev = math.sqrt(sum((x - mean)**2 for x in income_history) / len(income_history))
        cv      = std_dev / mean
        stability_score = max(0, 100 - cv * 200)
    else:
        stability_score = 100

    dti        = total_monthly_debt_payment / monthly_income
    debt_score = max(0, 100 - (dti / 0.5) * 100)

    insurance_score = 100 if has_insurance else 0

    final_score = (
        savings_score    * 0.35 +
        stability_score  * 0.25 +
        debt_score       * 0.25 +
        insurance_score  * 0.15
    )
    score = round(final_score)

    # ← English risk levels
    risk_level = (
        "SAFE"     if score >= 75 else
        "CAUTION"  if score >= 50 else
        "AT RISK"  if score >= 25 else
        "CRITICAL"
    )

    # ← English warning messages
    if savings_runway < 1:
        warning = f"Your savings only last {savings_runway * 30:.0f} days. This is an emergency."
    elif dti > 0.4:
        warning = f"{dti*100:.0f}% of your income goes to debt payments. Above safe limit."
    else:
        warning = f"You can survive {savings_runway:.1f} months without income."

    return {
        "score": score,
        "savings_runway_months": round(savings_runway, 1),
        "debt_to_income_ratio": round(dti * 100, 1),
        "risk_level": risk_level,
        "warning_message": warning,
        "breakdown": {
            "savings_score":   round(savings_score),
            "stability_score": round(stability_score),
            "debt_score":      round(debt_score),
            "insurance_score": round(insurance_score),
        }
    }